fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/ReorderJoinRule.java

```diff
@@ -177,23 +177,37 @@ public OutputColumnsPrune(OptimizerContext optimizerContext) {
             this.optimizerContext = optimizerContext;
         }
 
-        public OptExpression rewrite(OptExpression optExpression, ColumnRefSet pruneOutputColumns) {
+        public OptExpression rewrite(OptExpression optExpression, ColumnRefSet requiredColumns) {
             Operator operator = optExpression.getOp();
-            if (operator.getProjection() != null) {
+            if (operator.getProjection() != null && operator.getProjection().getColumnRefMap().size() > 1) {
                 Projection projection = operator.getProjection();
 
-                for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : projection.getColumnRefMap().entrySet()) {
-                    if (!entry.getValue().isColumnRef()) {
-                        return optExpression;
+                List<ColumnRefOperator> outputColumns = Lists.newArrayList();
+                for (ColumnRefOperator key : projection.getColumnRefMap().keySet()) {
+                    if (requiredColumns.contains(key)) {
+                        outputColumns.add(key);
                     }
+                }
 
-                    if (!entry.getKey().equals(entry.getValue())) {
-                        return optExpression;
+                if (outputColumns.size() == 0) {
+                    outputColumns.add(Utils.findSmallestColumnRef(projection.getOutputColumns()));
+                }
+
+                if (outputColumns.size() != projection.getColumnRefMap().size()) {
```

Comment:
The operator's output columns  will be pruned below, why set projection here

Reply:
Because projection put all join output columns(line:242), maybe contains some unnecessary columns, we need prune again,